src/Codegen/Constraints/BaseBuilder.php

```diff
@@ -92,6 +93,64 @@ protected function addEnumConstraintCheck(HackBuilder $hb): void {
     }
   }
 
+  protected function addHackEnumConstraintCheck(HackBuilder $hb): void {
+    $schema = type_assert_type($this->typed_schema, TSchema::class);
+    if (Shapes::keyExists($schema, 'hackEnum')) {
+      try {
+        $rc = new \ReflectionClass($schema['hackEnum']);
+      } catch (\ReflectionException $e) {
+        throw new \Exception(Str\format("Hack enum '%s' does not exist", $schema['hackEnum']));
+      }
+
+      invariant($rc->isEnum(), "'%s' is not an enum", $schema['hackEnum']);
+
+      $schema_type = $schema['type'] ?? null;
```

Comment:
I see, so the PR is extending both `integer` and `string` types values using `hackEnum` as the source for these values, rather than create a new `hackEnum` schema type first class citizen which I originally though was the case. This is the first time I see the repo so I don't know if creating a new type is easily doable so forgive me about some of the questions in advance.

If what I just mentioned is true, I understand we _need_ to specify the `type` because we are extending those 2 types. But if we were to define a new type, would using something like https://docs.hhvm.com/hack/reference/class/ReflectionClass/getEnumUnderlyingType/ help us validate the internal enum type if we say define a new
```
    "hack_enum": {
      "type": "hackEnum",
      "enum": "Slack\\Hack\\JsonSchema\\Tests\\TestStringEnum"
    }
```
?

Reply:
Specifying a new type would cause code generators for other languages (e.g., TypeScript) to break or to skip the property altogether, neither of which is desirable. Adding an extension to an existing type instead allows us to continue generating valid, reasonable schemas even when the generator doesn't understand the extension, which is the goal here.